architecture-docs/adr-003.md

```diff
@@ -0,0 +1,168 @@
+          unimplemented!("this feature is only used in merkle tree restore which we don't need yet");
+      }
+  }
+  ```
+
+- Implement `StakingGetter` with `JellyfishMerkleTree` against any `GetKV`,
+
+  Use `BlockHeight` as the version number.
+  
+  ```rust
+  pub struct StakingGetter<'a, S: GetKV> {
+      storage: &'a S,
+      block_height: BlockHeight,
+  }
+  
+  impl<'a, S: GetKV> Get for StakingGetter<'a, S> {
+      type Key = StakedStateAddress;
+      type Value = StakedState;
+      fn get(&self, key: &Self::Key) -> Option<Self::Value> {
+          JellyfishMerkleTree::new(&KVReader::new(self.storage))
+              .get_with_proof(HashValue::new(to_stake_key(key)), self.block_height.into())
+              .expect("merkle trie internal error")
+              .0
+              .map(|blob| {
+                  StakedState::decode(&mut blob.as_ref()).expect("merkle trie storage corrupted")
+              })
+      }
+  }
+  ```
+  
+- Use the same buffer abstraction as before:
+  
+  ```rust
+  pub type StakingBufferStore<'a, S, H> = BufferSimpleStore<'a, StakingGetter<'a, S>, H>;
+  pub type StakingBufferGetter<'a, S, H> = BufferGetter<'a, StakingGetter<'a, S>, H>;
+  ```
+  
+- When commit, flush the staking buffer into the Merkle trie, write the nodes into the `StoreKV`, which is the buffered
+  key-value storage shared with other states.
+  
+  ```rust
+  pub fn flush_stakings<S: StoreKV>(
+      storage: &mut S,
+      block_height: BlockHeight,
+      buffer: StakingBuffer,
+  ) -> Result<(usize, usize)> {
+      let reader = KVReader::new(storage);
+      let tree = JellyfishMerkleTree::new(&reader);
+      let (_, batch) = tree.put_blob_sets(
+          vec![buffer
+              .values()
+              .map(|staking| (HashValue::new(staking.key()), staking.encode().into()))
+              .collect::<Vec<_>>()],
+          block_height.into(),
+      )?;
+      for (key, node) in batch.node_batch.iter() {
+          storage.set((COL_TRIE_NODE, key.encode()?), node.encode()?);
+      }
+      storage.set(
+          (COL_STALED_NODE, block_height.encode()), 
+          batch.stale_node_index_batch.encode()
+      );
+      Ok((batch.num_new_leaves, batch.num_stale_leaves))
+  }
+  ```
+  
+- A complete commit would be like this:
+  
+  ```rust
+  flush_staking(&mut kv_store!(self), block_height, staking_buffer);
+  flush_kv(&mut self.kvdb, kv_buffer);  // Write to the disk atomically.
+  ```
+
+### Chain node state changes
+
+- Remove `AccountStorage` from `ChainNodeApp`.
+
+- In `ChainState`, change `account_root` to `last_account_block_height`, which records the last block height there's any
+  staking state modified.
+  
+  Use the block height to query history state directly.
+
+### Staled nodes cleanup strategy
+
+- The keys of staled nodes are stored in column `COL_STALED_NODE`, indexed by block height.
+- Periodically cleanup the nodes for block heights too old by iterate keys in colume `COL_STALED_NODE`.
```

Comment:
What are "staled nodes"?

Reply:
> What are "staled nodes"?

The nodes not used in newer block heights of the tree, only referenced by old block heights.
It's possible to clean them up, if we don't want to keep all the old histories.

Reply:
it'll be good to explain this in the doc -- with starling, I think it's also possible to prune unused nodes, but it wasn't done.
It'll make sense to do this in the "validator" setup (where one just cares about the latest state in order to validate new transactions), but for "full node" setup, one will want to keep them, as one may need to query a historical state with a proof.

Not sure how much extra complexity it is -- one particular thing here would be whether "integration test' would be possible (to test that the cleanup mechanism doesn't accidentally cause problems)... perhaps it may be easier to leave it out first? The setting for cleanup can be in a different PR?

Reply:
I'll mark it as a potencial feature for the future, we keep all the old states for now.